File: 07 - Multirate Digital Filtering/matlab/ex02.m
%% Target
%
% Design a fractional rate converter and use it to filter a signal.
% 
% The parameters are:
%
% * x[n] = cos(2*pi*Fc*n*Ts) + w[n];
% * w[n] is the noise
% * Fc = 62.5 [kHz]
% * F  =  1   [MHz]
% * L  =  5   % Interpolator Factor
% * M  =  3   % Decimator Factor
% 
% *Suggestions*
% 
% FIR coefficients
%   fir_ord = 64;
%   [fir_num] = fir1(fir_ord, Wn, hann(fir_ord+1));
% 
% Filtering
%   y = filter(num,den,x);
%     * x   - input signal
%     * y   - output signal
%     * num - numerator   coefficients
%     * den - denominator coefficients
%

%% Clear
clc;            % Clear the text from the Command Window
clear;          % Remove all variables from the current workspace
close all;      % Close all figures

%% Matlab Exercise
Fc = 125e3;
Fs = 1e6;
Ts = 1/Fs;
len = 2^12;
L = 5;
M = 3;
n = 0:len-1;

% Signal
w = 0.01*randn(1,len);
x = cos(2*pi*Fc*n*Ts) + w;

% Filter
fir_ord = 64;
F_cut = 500e3;
Wn = F_cut / (Fs*L/2);
fir_num = fir1(fir_ord, Wn, hann(fir_ord+1));

% Upsampler
xu = upsample(x,L);
nu = 0:length(xu)-1;

% Interpolation
y = L*filter(fir_num, 1, xu);
y(1:fir_ord/2) = [];          % Remove the filter latency

% Downsampler
y = downsample(y,M);
ny = 0:length(y)-1;

% Time vectors
tx  = n*Ts / 1e6;
ty  = ny*(Ts/L*M) / 1e6;

%% Plot
% Using 'freqz'
nFFT = 2^12; % Number of points of the fft
Hf_fir  = freqz(fir_num, 1, nFFT);
Xf      = freqz(x, 1, nFFT);
[Yf, w] = freqz(y, 1, nFFT);

% Frequency normalization
w = w/pi * (Fs/2)/1e6;
w_u = w*L;
w_y = w*L/M;

% Mag to dB
Hf_fir = mag2db(abs(Hf_fir));
Xf     = mag2db(abs(Xf)/nFFT);
Yf     = mag2db(abs(Yf)/nFFT);

figure
subplot(3,1,1)
  hold on
  plot(w_u, Hf_fir)
  hold off
  grid on
  legend('H_{FIR}')
  xlabel('Frequency [MHz]')
  ylabel('Amplitude [dB]')
subplot(3,1,2)
  hold on
  plot(w,   Xf,   's-')
  plot(w_y, Yf,   'x-')
  hold off
  grid on
  legend('x','y')
  xlabel('Frequency [MHz]')
  ylabel('Amplitude [dB]')
subplot(3,1,3)
  hold on
  plot(tx ,  x, 's-')
  plot(ty ,  y, 'x-')
  hold off
  grid on
  legend('x','y')
  xlabel('Time \mus')
  ylabel('Amplitude')

% Using 'fvtool'
hfvt0 = fvtool(x,  1, 'Fs', Fs);
legend(hfvt0, 'x')
hfvt1 = fvtool(fir_num, 1, 'Fs', Fs*L);
legend(hfvt1, 'FIR')
hfvt2 = fvtool(y,  1, 'Fs', Fs*L/M);
legend(hfvt2, 'y')

File: 07 - Multirate Digital Filtering/python/ex02.py
## Target
#
# Design an interpolator and use it to filter a signal.
# 
# The parameters are:
#
# * x[n] = cos(2*pi*Fc*n*Ts) + w[n];
# * w[n] is the noise
# * Fc = 125 [kHz]
# * Fs =   1 [MHz]
# * L  =   4  # Interpolator Factor
# 

## Import libraries
import numpy as np
import matplotlib.pyplot as plt

# Import the submodule signal of scipy to use the correlation funciton
from scipy import signal

## Parameters
Fc = 125e3
Fs = 1e6
Ts = 1.0/Fs
len = 2**12
L = 5
M = 3

# Index vector of the samples
n  = np.arange(start=0, stop=len, dtype=int)

# Signal
mu = 0
sigma = 1e-2
w = np.random.normal(mu, sigma, size=(len))
x = np.cos(2*np.pi*Fc*n*Ts) + w

# FIR Filter
fir_ord = 64
fir_len = fir_ord+1
F_cut = 500e3
Wn = F_cut / (Fs*L/2.0);
fir_num = signal.firwin(numtaps=fir_len,
                        cutoff=Wn, 
                        window='hann')

# Upsample
xu = np.zeros(L*len, dtype='float')
xu[::L] = x
nu = np.arange(start=0, stop=np.size(xu), dtype=int)

# Filtering
y = L*signal.lfilter(b=fir_num, a=1, x=xu)

# Remove the filter latency
y = y[fir_ord/2:-fir_ord/2]

# Downsample
y = y[::M]

# Time vectors
ny = np.linspace(start=0, stop=np.size(y)-1, num=np.size(y))
tx  = n*Ts * 1e6
ty  = ny*(Ts/L*M) * 1e6

## Figure

# Frequency vector for the plot
nFFT = 2**12
w, Hf_fir = signal.freqz(b=fir_num, a=1,
                         worN=nFFT,
                         whole=False)
w, Xf = signal.freqz(b=x,a=np.size(x),
                    worN=nFFT,
                    whole=False)
w, Yf = signal.freqz(b=y,a=np.size(y),
                    worN=nFFT,
                    whole=False)
# Frequency normalization
w = w/np.pi * (Fs/2)/1e6
w_u = w*L
w_y = w*L/M

# Mag to dB
Hf_fir = 10*np.log10(np.abs(Hf_fir)**2)
Xf = 10*np.log10(np.abs(Xf)**2)
Yf = 10*np.log10(np.abs(Yf)**2)


plt.figure()

plt.subplot(3,1,1)
plt.plot(w, Hf_fir, '-', label='H_FIR')
plt.grid()
plt.legend(loc='upper right')
plt.xlabel('Frequency [MHz]')
plt.ylabel('Amplitude [dB]')

plt.subplot(3,1,2)
plt.plot(w,   Xf,   '-', label='x')
plt.plot(w_y, Yf,   '-', label='y')
plt.grid()
plt.legend(loc='upper right')
plt.xlabel('Frequency [MHz]')
plt.ylabel('Amplitude [dB]')

plt.subplot(3,1,3)
plt.plot(tx, x,   's-', label='x')
plt.plot(ty, y,   'x-', label='y')
plt.grid()
plt.legend(loc='upper right')
plt.xlabel('Time \mu s')
plt.ylabel('Amplitude')

plt.show()
